src/inspector.rs

```diff
@@ -0,0 +1,675 @@
+/*
+ * Copyright Redis Ltd. 2022 - present
+ * Licensed under your choice of the Redis Source Available License 2.0 (RSALv2) or
+ * the Server Side Public License v1 (SSPLv1).
+ */
+//! The [Inspector] is a helpful facility for remote debugging of the
+//! JavaScript code provided by the V8 engine.
+//!
+//! To be able to debug, an inspector must be created and supplied with
+//! the remote debugger's requests. To answer back, the `on_response`
+//! callback must be properly set.
+//!
+//! # Sessions
+//!
+//! To be able to debug remotely, a server is required. Here one may
+//! find the [WebSocketServer] useful. The [WebSocketServer] is a very
+//! simple WebSocket server, which can be used with an [Inspector].
+//!
+//! To start a debugging server session, it is required to have a
+//! [V8ContextScope], a [WebSocketServer] and an [Inspector]. All of
+//! those are gently packed within an easy-to-use [DebuggerSession]
+//! struct which abstracts all the work. It is a scoped object, meaning
+//! that once an object of [DebuggerSession] leaves its scope, the
+//! debugger server and the debugging session are stopped.
+//!
+//! # Clients
+//!
+//! To connect to a remote debugging WebSocker server, one can use:
+//!
+//! 1. A custom web-socket client, and some implementation of the
+//! V8 Inspector protocol.
+//! 2. The chrome-devtools, for example, by running the chromium
+//! web-browser and navigating to
+//! <devtools://devtools/bundled/inspector.html?ws=ip:port> where
+//! the `ip`/`port` parameters are the corresponding ip address and port
+//! of the [WebSocketServer] used.
+//! 3. (Needs verification) Using the WebKit MiniBrowser
+//! <https://trac.webkit.org/wiki/RemoteInspectorGTKandWPE>.
+//! 4. Using the V8's [D8 console debugger](https://v8.dev/docs/d8).
+//! 5. Using the [Visual Studio Code](https://code.visualstudio.com/),
+//! for example, using a configuration like this:
+//!     ```json
+//!     {
+//!         "version": "0.2.0",
+//!         "configurations": [
+//!             {
+//!                 "name": "Attach to RedisGears V8 on port 9005",
+//!                 "type": "node",
+//!                 "request": "attach",
+//!                 "cwd": "${workspaceFolder}",
+//!                 "websocketAddress": "ws://127.0.0.1:9005",
+//!             }
+//!         ]
+//!     }
+//!     ```
+//!
+use std::net::TcpListener;
+use std::rc::Rc;
+use std::sync::Mutex;
+use std::time::Duration;
+
+use serde::Deserialize;
+
+use crate::v8::v8_context_scope::V8ContextScope;
+
+/// The remote debugging server port for the [WebSocketServer].
```

Comment:
I wonder if this code should be on the crate user side, and this crate should only expose the inspector API?